File: src/swift-camera/RealTimePPGSignalsProcessor.swift
import Foundation
import AVFoundation
import CoreMedia
import CoreVideo

// Swift equivalent of RealTimePPGSignalsProcessor.cs
class RealTimePPGSignalsProcessor: NSObject, AVCaptureVideoDataOutputSampleBufferDelegate {
    private let ppgSignalsProcessor: PPGSignalsProcessor
    private var freeFrames: [Frame] = []
    private let freeFramesLock = NSLock()
    
    init(ppgSignalsCollector: IPPGSignalsCollector) {
        self.ppgSignalsProcessor = PPGSignalsProcessor(ppgSignalsCollector: ppgSignalsCollector, freeFrames: freeFrames)
        super.init()
    }
    
    func terminateProcessing() {
        ppgSignalsProcessor.terminateProcessing()
    }
    
    func processingToCompletionAsync() async {
        await ppgSignalsProcessor.processingToCompletionAsync()
    }
    
    // MARK: - AVCaptureVideoDataOutputSampleBufferDelegate
    
    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        do {
            let frame = bufferToFrame(sampleBuffer)
            ppgSignalsProcessor.enqueue(frame)
        } catch {
            print("Error processing sample buffer: \(error)")
        }
    }
    
    private func bufferToFrame(_ sampleBuffer: CMSampleBuffer) -> Frame {
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            fatalError("Could not get pixel buffer from sample buffer")
        }
        
        CVPixelBufferLockBaseAddress(pixelBuffer, CVPixelBufferLockFlags(rawValue: 0))
        defer {
            CVPixelBufferUnlockBaseAddress(pixelBuffer, CVPixelBufferLockFlags(rawValue: 0))
        }
        
        // Try to get a free frame from the pool
        freeFramesLock.lock()
        let frame = freeFrames.isEmpty ? Frame(Y: Data(), U: Data(), V: Data(), UV: Data(), timestamp: 0) : freeFrames.removeLast()
        freeFramesLock.unlock()
        
        // Set timestamp
        let presentationTimeStamp = CMSampleBufferGetPresentationTimeStamp(sampleBuffer)
        frame.timestamp = UInt64(presentationTimeStamp.value)
        
        // Extract Y plane
        let yWidth = CVPixelBufferGetWidthOfPlane(pixelBuffer, 0)
        let yHeight = CVPixelBufferGetHeightOfPlane(pixelBuffer, 0)
        let ySize = yWidth * yHeight
        if let yAddress = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 0) {
            frame.Y = Data(bytes: yAddress, count: ySize)
        }
        
        // Extract UV plane
        let uvHeight = CVPixelBufferGetHeightOfPlane(pixelBuffer, 1)
        let uvBytesPerRow = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 1)
        let uvSize = uvHeight * uvBytesPerRow
        if let uvAddress = CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 1) {
            frame.UV = Data(bytes: uvAddress, count: uvSize)
        }
        
        return frame
    }
}

File: src/original-c#-camera/RealTimePPGSignalsProcessor.cs
using AVFoundation;
using CoreMedia;
using CoreVideo;
using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Threading.Tasks;

namespace RivaDigitalApp.iOS
{
    public class RealTimePPGSignalsProcessor : AVCaptureVideoDataOutputSampleBufferDelegate
    {
        private readonly PPGSignalsProcessor _ppgSignalsProcessor;
        private readonly ConcurrentStack<Frame> _freeFrames = new ConcurrentStack<Frame>();

        public RealTimePPGSignalsProcessor(IPPGSignalsCollector ppgSignalsCollector)
        {
            _ppgSignalsProcessor = new PPGSignalsProcessor(ppgSignalsCollector, _freeFrames);
        }

        public void TerminateProcessing()
        {
            _ppgSignalsProcessor.TerminateProcessing();
        }

        public Task ProcessingToCompletionAsync()
        {
            return _ppgSignalsProcessor.ProcessingToCompletionAsync();
        }

        public override void DidOutputSampleBuffer(AVCaptureOutput captureOutput, CMSampleBuffer sampleBuffer, AVCaptureConnection connection)
        {
            try
            {
                var frame = BufferToFrame(sampleBuffer);
                _ppgSignalsProcessor.Enqueue(frame);
            }
            catch (Exception e)
            {
                Debug.WriteLine(e);
            }
            finally
            {
                sampleBuffer.Dispose();
            }
        }

        private Frame BufferToFrame(CMSampleBuffer sampleBuffer)
        {
            using (var pixelBuffer = sampleBuffer.GetImageBuffer() as CVPixelBuffer)
            {
                pixelBuffer.Lock(CVPixelBufferLock.None);

                try
                {
                    if (!_freeFrames.TryPop(out Frame frame))
                    {
                        frame = new Frame();
                    }

                    frame.Timestamp = (ulong)sampleBuffer.PresentationTimeStamp.Value;

                    var yWidth = pixelBuffer.GetWidthOfPlane(0);
                    var yHeight = pixelBuffer.GetHeightOfPlane(0);
                    int ySize = (int)(yWidth * yHeight);
                    var yAddress = pixelBuffer.GetBaseAddress(0);
                    UpdateBuffer(ref frame.Y, ySize, yAddress);

                    var uvHeight = pixelBuffer.GetHeightOfPlane(1);
                    var uvBytesPerRow = pixelBuffer.GetBytesPerRowOfPlane(1);
                    int uvSize = (int)(uvHeight * uvBytesPerRow);
                    var uvAddress = pixelBuffer.GetBaseAddress(1);
                    UpdateBuffer(ref frame.UV, uvSize, uvAddress);

                    return frame;
                }
                finally
                {
                    pixelBuffer.Unlock(CVPixelBufferLock.None);
                }

                void UpdateBuffer(ref byte[] buffer, int size, IntPtr sourceAddress)
                {
                    if (buffer?.Length != size)
                    {
                        buffer = new byte[size];
                    }
                    Marshal.Copy(sourceAddress, buffer, 0, size);
                }
            }
        }
    }
}